compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateAIEWorkgroup.h

```diff
@@ -0,0 +1,176 @@
+    assert(source->getNumArguments() == target->getNumArguments() &&
+           "expected same number of arguments");
+    for (auto &&[sourceArg, targetArg] :
+         llvm::zip(source->getArguments(), target->getArguments())) {
+      mapper.map(sourceArg, targetArg);
+    }
+  }
+
+  /// The map to be used for mapping and looking up IR entities.
+  IRMapping mapper;
+};
+
+/// Utility class to contain and maintain the core operations' as a map from
+/// coordinates to the respective core operation on that location. The core map
+/// can be accessed through lookup functions and new entries can be added
+/// through the 'mapOrMerge' method or by merging with another CoreContext. This
+/// is useful for creating a clean context for building nested operations and
+/// then merging with the outer context.
+class CoreContext {
+ public:
+  CoreContext(IRRewriterAndMapper &rewriter) : rewriter(rewriter) {}
+  CoreContext(IRRewriterAndMapper &&rewriter) = delete;
+
+  // Check whether the coordinate exists in the map.
+  bool contains(const std::tuple<int64_t, int64_t> &coordinate) {
+    return coreMap.contains(coordinate);
+  }
+
+  /// Return the underlying core map.
+  DenseMap<std::tuple<int64_t, int64_t>, AMDAIE::CoreOp> &getCoreMap() {
+    return coreMap;
+  }
+
+  /// Lookup a coordinate in the core map. This asserts that the provided
+  /// coordinate exists.
+  AMDAIE::CoreOp lookup(const std::tuple<int64_t, int64_t> &coordinate) {
+    AMDAIE::CoreOp res = lookupOrNull(coordinate);
+    assert(res && "expected 'coordinate' to be found in the map");
+    return res;
+  }
+
+  /// Return the underlying core map.
+  AMDAIE::CoreOp lookupOrNull(const std::tuple<int64_t, int64_t> &coordinate) {
+    return contains(coordinate) ? coreMap[coordinate] : AMDAIE::CoreOp(nullptr);
+  }
+
+  // Inserts a new mapping from 'coordinate' to 'coreOp' or merges with a
+  // potentially existing entry.
+  void mapOrMerge(const std::tuple<int64_t, int64_t> &coordinate,
+                  AMDAIE::CoreOp coreOp) {
+    AMDAIE::CoreOp existingCoreOp = lookupOrNull(coordinate);
+    if (!existingCoreOp) {
+      coreMap[coordinate] = coreOp;
+    } else {
+      mergeCoreOps(coreOp, existingCoreOp);
+    }
+  }
+
+  /// Merge another context with this one.
+  void mergeContext(CoreContext &other) {
+    for (auto &&[coordinate, coreOp] : other.getCoreMap())
+      mapOrMerge(coordinate, coreOp);
+  }
+
+ private:
+  /// Merge the 'source' core operations in the end of the 'dest' core
+  /// operation.
+  void mergeCoreOps(AMDAIE::CoreOp source, AMDAIE::CoreOp dest);
+
+  /// The rewriter to be used.
+  IRRewriterAndMapper &rewriter;
+
+  /// Map from coordinates, represented as a tuple of integers, to the
+  /// respective core operation on that location.
+  DenseMap<std::tuple<int64_t, int64_t>, AMDAIE::CoreOp> coreMap;
+};
+
+/// Recursive workgroup build function for an operation.
+LogicalResult workgroupBuild(IRRewriterAndMapper &rewriter, Operation *op,
```

Comment:
Please check these recursive functions. Is there existing infrastructure I can use to accomplish the same?

Reply:
None that I'm aware of with the degree of control needed for individual operations.

CC: @MaheshRavishankar could you check ?